zaza/openstack/charm_tests/keystone/tests.py

```diff
@@ -371,3 +372,87 @@ def test_security_checklist(self):
             expected_passes,
             expected_failures,
             expected_to_pass=False)
+
+
+class LdapTests(BaseKeystoneTest):
+    """Keystone ldap tests tests."""
+
+    @classmethod
+    def setUpClass(cls):
+        """Run class setup for running Keystone ldap-tests."""
+        super(LdapTests, cls).setUpClass()
+
+    def _get_ldap_config(self):
+        ldap_ips = zaza.model.get_app_ips("ldap-server")
+        ldap_server_ip = ldap_ips[0]
+
+        keystone_ldap_config = {
+            'ldap-server': "ldap://{}".format(ldap_server_ip),
+            'ldap-user': 'cn=admin,dc=test,dc=com',
+            'ldap-password': 'crapper',
+            'ldap-suffix': 'dc=test,dc=com',
+            'domain-name': 'userdomain',
+        }
+        if all(keystone_ldap_config.values()):
+            self.ldap_configured = True
+            return keystone_ldap_config
+        else:
+            # NOTE(jamespage): Use mock values to check deployment only
+            #                  as no test fixture has been supplied
+            self.ldap_configured = False
+            return {
+                'ldap-server': 'myserver',
+                'ldap-user': 'myuser',
+                'ldap-password': 'mypassword',
+                'ldap-suffix': 'mysuffix',
+                'domain-name': 'userdomain',
+            }
+
+    # NOTE: intermittent authentication fails. Wrap in a retry loop.
+    @tenacity.retry(wait=tenacity.wait_exponential(multiplier=1,
+                                                   min=5, max=10),
+                    reraise=True)
+    def _find_keystone_v3_user(self, username, domain):
+        """Find a user within a specified keystone v3 domain."""
+        client = self.admin_keystone_client
+        domain_users = client.users.list(
+            domain=client.domains.find(name=domain).id
+        )
+        usernames = []
+        for user in domain_users:
+            usernames.append(user.name)
+            if username.lower() == user.name.lower():
+                return user
+        logging.debug("User {} was not in these users: {}. Returning None."
+                      "".format(username, usernames))
+        return None
+
+    def test_100_keystone_ldap_users(self):
+        """Validate basic functionality of keystone API with ldap."""
+        zaza.model.set_application_config('keystone-ldap',
+                                          self._get_ldap_config())
+
+        if not self.ldap_configured:
```

Comment:
It would be better if `_get_ldap_config()` returned a tuple of `(configured, config_data)` or raises an Exception if the LDAP couldn't be configured.  By setting an instance var in the method, this is effectively a side-effect leak which is much harder to refactor and reason about.

e.g.
```python
(configured, ldap_config) = self._get_ldap_config()
if not configured:
    raise unittest.SkipTest(...)
# now do config
with config_change(...):
     # now do the  checks
```

This way you can avoid starting the config change before you have the config (rather than setting it to `None`) and it's more obvious *where* self.ldap_configured is actually coming from, as it a return value from the function rather than an object instance var.

The alternative is to raise the `SkipTest` *in* `_get_ldap_config()` which would skip the test and bail early; however, I tend to prefer a more explicit approach.